src/shell_manager.cc

```diff
@@ -325,6 +325,232 @@ std::pair<String, int> ShellManager::eval(
+                                    watcher.disable();
+                                    return;
+                                }
+                            }
+                        });
+            int flags = fcntl(this->pipe.write_fd(), F_GETFL, 0);
+            fcntl(this->pipe.write_fd(), F_SETFL, flags | O_NONBLOCK);
+        }
+    };    
+
+    struct PipeProcess
+    {
+        pid_t pid;
+        PipeWriter &&stdin;
+        PipeReader &&stdout, &&stderr;
+    };
+
+    struct PipeProcessContainer {
+        Vector<PipeProcess> processes;
+
+        std::pair<bool,int> is_terminated() {
+            bool terminated = true;
+            int status = 0;
+            for (PipeProcess& process: processes) {
+                terminated = terminated && (waitpid(process.pid, &status, WNOHANG) != 0);
+            }
+            return {terminated,status};
+        }
+
+        bool writes_closed() {
+            return std::none_of(processes.cbegin(),processes.cend(),
+                 [](const PipeProcess& p) {
+                     return p.stdin.pipe.write_fd() != -1;
+                 });
+        }
+
+        bool reads_closed() {
+            return std::none_of(processes.cbegin(),processes.cend(),
+                 [](const PipeProcess& p) {
+                     return p.stdout.pipe.read_fd() != -1 
+                        and p.stderr.pipe.read_fd() != -1;
+                 });
+        }
+    } process_container;
+
+    auto wait_time = Clock::now();
+
+    for (StringView input: inputs) 
+    {
+        Pipe child_stdin{not input.empty()}, child_stdout, child_stderr;
+        pid_t pid = spawn_shell(m_shell.c_str(), cmdline, shell_context.params, kak_env,
+                                [&child_stdin, &child_stdout, &child_stderr] {
+            auto move = [](int oldfd, int newfd)
+            {
+                if (oldfd == newfd)
+                    return;
+                dup2(oldfd, newfd); close(oldfd);
+            };
+
+            if (child_stdin.read_fd() != -1)
+            {
+                close(child_stdin.write_fd());
+                move(child_stdin.read_fd(), 0);
+            }
+            else
+                move(open("/dev/null", O_RDONLY), 0);
+
+            close(child_stdout.read_fd());
+            move(child_stdout.write_fd(), 1);
+
+            close(child_stderr.read_fd());
+            move(child_stderr.write_fd(), 2);
+        });
+
+        child_stdin.close_read_fd();
+        child_stdout.close_write_fd();
+        child_stderr.close_write_fd();
+
+        String stdout_contents, stderr_contents;
+        PipeReader stdout_reader{child_stdout, stdout_contents};
+        PipeReader stderr_reader{child_stderr, stderr_contents};
+        PipeWriter stdin_writer{child_stdin, input};
+        process_container.processes.push_back(PipeProcess{
+            pid,
+            std::move(stdin_writer),
+            std::move(stdout_reader),std::move(stderr_reader)
```

Comment:
You are storing references to local objects that are going to die immediatly when you hit the closing curly brace of that loop. Additionally, their storage on the stack is very likely to be re-used by the same variables for the next loop iteration, which will lead to all the processes in the process container to reference the same (dead) PipeReaders and PipeWriter...

Reply:
As stated above, the goal was to actually move the value to be stored inside the `PipeProcess` struct (to avoid precisely the problem of the values going out of scope), but apparently this is not the correct syntax/semantic for that. 
Your explanation would explain the behaviour I'm noticing